### DISCUSSION TITLE
is it possible to obtain adjoint surface sensitivities w.r.t surface mesh nodes?

### DISCUSSION THREAD

**Discussion by Mohsin-Bashir** [mdolab/dafoam#544] [General]
1. For example lets take the case of NACA0012/OneraM6, if i want to calculate the adjoint surface sensitivities w.r.t each of its surface mesh node. **Is it possible?** 

2. In FAQ "https://dafoam.github.io/mydoc_get_started_faq.html#how-to-get-sensitivity-maps", it says "Now, we only support outputting sensitivity for FFD and Field designVarType"...
Here what does it mean by Field variable?

From this what I understood is if I change task name in runScript_v2.py from "opt" to "runAdjoint" and this line 
`"writeSensMap": [ "shapey"],`
it will output a new file name "totalDerivHist.txt" which I guess contains total derivatives w.r.t. each shapey (for NACA0012 case). 
I have already read this thread "https://github.com/mdolab/dafoam/discussions/341", here it says "It will compute the sens map for surface nodes, not FFD." but where will it save the corresponding file and with what name?

3. how can I view sensitivity maps in paraview, which file do I need to load in paraview?

**Comment by Mohsin-Bashir** [comment]
Regarding the second question, I have found sensitivity files here `incompressible\processor0\0.00000000001` with name "sens_CD_shapey". Just need to look for sensitivity files in different processor folders.

But I still don't get it which sensitivities are corresponding to which node.

**Comment by friedenhe** [comment]
The sensitivity field/file is ordered the same way as other variables, such as U, p, and nut. So you can visualize the surface sensitivity in Paraview.

**Reply by Mohsin-Bashir** [reply]
Thanks for your response, I get it that all variables are ordered in same way. 
e.g some of the surface sensitivities are as below
```
(
(-0.000077 0.000713 0.000861)
(-0.000106 0.000961 0.000909)
(-0.000124 0.001096 0.000948)
(-0.000153 0.001329 0.000980)
(-0.000192 0.001628 0.000986)
(-0.000251 0.002066 0.000984)
(-0.000294 0.002363 0.000987)
(-0.000378 0.002971 0.000969)
(-0.000506 0.003899 0.000945)
(-0.000639 0.004833 0.000832)
(-0.000746 0.005549 0.000630)
(-0.000908 0.006639 0.000451)
(-0.001201 0.008690 0.000307)
(-0.001282 0.009120 0.000193)
(-0.001477 0.010365 0.000106)
(-0.001944 0.013571 0.000062)
)
```
But I want to know if there is a way to find out their surface mesh coordinates in the same order.

**Reply by Mohsin-Bashir** [reply]
![image](https://github.com/mdolab/dafoam/assets/118594705/8a465aa4-26f7-414b-b396-5dd4a0bce9d7)
Here in paraview surface sensitivity is ordered against surface mesh node coordinates. I want this information in readable file like .txt etc. 
I need to use this information in another code, but I don't know if its possible to get this in readable file. (I don't want to open paraview and export this unless there is some automated way to do this) Any help regarding this will be much appreciated.

**Comment by Mohsin-Bashir** [comment]
I was reading one of your paper https://doi.org/10.1016/j.paerosci.2019.05.002, I just want to confirm if this is the surface sensitivity term (Circled in picture below)?
![Capturstae](https://github.com/mdolab/dafoam/assets/118594705/69f2ec2c-69dd-44f7-a330-a801aaaf4c14)

**Comment by friedenhe** [comment]
> But I want to know if there is a way to find out their surface mesh coordinates in the same order.

Please use the mesh.Cf().boundaryField()[patchI][faceI] to get the surface center coordiates for a give patch number (patchI) and face number (faceI). Please google how mesh.Cf() works in OpenFOAM

**Comment by friedenhe** [comment]
> I was reading one of your paper https://doi.org/10.1016/j.paerosci.2019.05.002, I just want to confirm if this is the surface sensitivity term (Circled in picture below)?  

Yes
